### PR TITLE
[issue 7293] add test for rettangle box decoration with boarder radius

### PR DESCRIPTION
**Pull request by wise86-android**
## Description

we increased the number of tests to check the decoration box that the tap has a consistent behavior when the decoration is a rectangle with a rounded corner.
We test it doing a tap just outside the corner and just inside. 

this code was made during an  [Open Source Saturday](https://www.meetup.com/it-IT/Open-Source-Saturday-Milano/)

## Related Issues

https://github.com/flutter/flutter/issues/7293

## Tests

I added the following tests:
 - BoxDecoration not tap outside rounded angles - Top Left
 - BoxDecoration tap inside rounded angles - Top left
 - BoxDecoration rounded angles other corner works

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 1]
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:math';
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: flutter style for math is `import 'dart:math' as math`.

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 2]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
 
   });
 
+  testWidgets('Can hit test on BoxDecoration border', (WidgetTester tester) async {
+
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: extra new line

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 3]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
...
+    await tester.tapAt(const Offset(350.0, 275.0));
+    expect(itemsTapped, isEmpty);
+
+    await tester.tapAt(const Offset(400.0, 300.0));
+    expect(itemsTapped, <int>[1]);
+
+    await tester.tap(find.byKey(key));
+    expect(itemsTapped, <int>[1,1]);
+
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: extra new line

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 4]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
...
+    const Offset borderTopTangent = Offset(radius-1, 0.0);
+    const Offset borderLeftTangent = Offset(0.0,radius-1);
+    //the borderDiagonalOffset is the backslash line
+    //\\######@@@
+    //#\\###@####
+    //##\\@######
+    //##@########
+    //@##########
+    //@##########
+    const double borderDiagonalOffset = radius - radius*sqrt1_2;
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: `radius - radius * math.sqrt1_2` (spaces)

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 5]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
...
+      );
+    }
+
+    await tester.pumpWidget(buildFrame(Border.all()));
+
+    expect(itemsTapped, isEmpty);
+    // x, y
+    const Offset borderTopTangent = Offset(radius, 0.0);
+    const Offset borderLeftTangent = Offset(0.0,radius);
+    const double borderDiagonalOffset = radius - radius*sqrt1_2;
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: ditto

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 6]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
...
+
+    await tester.tapAt(borderLeftTangent);
+    expect(itemsTapped, <int>[1,1],reason: 'border Left not tapped');
+
+    await tester.tapAt(fartherBorderRadiusPoint);
+    expect(itemsTapped, <int>[1,1,1],reason: 'border center not tapped');
+
+    await tester.tap(find.byKey(key));
+    expect(itemsTapped, <int>[1,1,1,1]);
+
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: blank line

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 7]
```diff
@@ -342,4 +343,221 @@ Future<void> main() async {
...
+    await tester.tapAt(bottomLeftOutside);
+    expect(itemsTapped, <int>[1,1,1,1],reason: 'bottom left tapped');
+
+    await tester.tapAt(topLeftInside);
+    expect(itemsTapped, <int>[1,1,1,1,1],reason: 'top left not tapped');
+
+    await tester.tapAt(topLeftOutside);
+    expect(itemsTapped, <int>[1,1,1,1,1],reason: 'top left tapped');
+  });
+
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: nit: blank line

**Review by dnfield — APPROVED** [review]
LGTM modulo some style nits.  Thanks!

**Comment by wise86-android** [comment]
Hi Dan,

I run flutter format on the file, and use the flutter style to import math lib.

**Comment by dnfield** [comment]
Unfortunately, we don't use dartfmt on the source in this repo - it's all hand formatted.  

Can you revert the new formatting and just  fix the issues inponted out?  Sorry for the inconveniences!

**Comment by goderbauer** [comment]
@wise86-android Do you have plans to come back to this PR soon to address the comments?

File: packages/flutter/test/widgets/box_decoration_test.dart [hunk 8]
```diff
@@ -342,4 +343,218 @@ Future<void> main() async {
...
+}
```
Comment (dnfield) [inline packages/flutter/test/widgets/box_decoration_test.dart]: Please remove the extra blank line and the new line at EOF - then we should be good :)

Comment (wise86-android) [reply packages/flutter/test/widgets/box_decoration_test.dart]: done, you should think something for add an automatic formatting to the repository..

Comment (dnfield) [reply packages/flutter/test/widgets/box_decoration_test.dart]: I know it's a bit of a pain, sorry about that.  You can read more about the rationale at https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#in-defense-of-the-extra-work-that-hand-formatting-entails

**Review by dnfield — APPROVED** [review]
LGTM, thanks for sticking with this!

**Pull request merged**